src/precice/impl/Participant.hpp

```diff
@@ -47,7 +48,16 @@ namespace impl {
 
 /// Holds coupling state of one participating solver in coupled simulation.
 class Participant {
+private:
+  // needs to be declared here for typedef to work properly
+  std::map<DataID, WriteDataContext> _writeDataContexts;
+
+  std::map<DataID, ReadDataContext> _readDataContexts;
```

Comment:
I don't like this "forward declaration" so much. I tried to implement something along the following lines yesterday, but unfortunately did not get it working due to a `const` somewhere in the generated types:
```
class Participant {
public:
  using ReadDataContextMap = std::map<DataID, ReadDataContext>;
  using WriteDataContextMap = std::map<DataID, WriteDataContext>;
  
  using ReadDataContextValues = decltype(std::declval<ReadDataContextMap>() | boost::adaptors::map_values);
  using WriteDataContextValues = decltype(std::declval<WriteDataContextMap>() | boost::adaptors::map_values);
  ...
  ReadDataContextValues readDataContexts()
  {
    return _readDataContexts | boost::adaptors::map_values;
  }

  WriteDataContextValues writeDataContexts()
  {
    return _writeDataContexts | boost::adaptors::map_values;
  }
  ...
private:
  ReadDataContextMap _readDataContexts;
  
  WriteDataContextMap _writeDataContexts;
  ...
};
```
@fsimonis any C++ insights would be highly appreciated ;)

Reply:
Let's reduce the pain by moving the implementation into the header.

This allows you to write:
```cpp
auto writeDataContexts() {
  return _writeDataContexts | boost::adaptors::map_values;
}
```

This way, you don't have to move the declaration nor add a type alias for the values.

Reply:
8292a6e done.